core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -941,14 +942,30 @@ boolean canRemoveCastFromLiteral(RelDataType toType,
   private RexNode makeCastExactToBoolean(RelDataType toType, RexNode exp) {
     return makeCall(toType,
         SqlStdOperatorTable.NOT_EQUALS,
-        ImmutableList.of(exp, makeZeroLiteral(exp.getType())));
+        ImmutableList.of(exp, makeZeroValue(exp.getType())));
+  }
+
+  /** Some data types do not have literals; this creates an expression that
+   * evaluates to a zero of the specified type.
+   *
+   * @param type A numeric type.
+   * @return     An expression that evaluates to 0 of the specified type.
+   */
+  public RexNode makeZeroValue(RelDataType type) {
+    if (SqlTypeUtil.hasLiterals(type)) {
+      return makeZeroLiteral(type);
+    } else {
+      // This is e.g., an unsigned type
+      RelDataType i = typeFactory.createSqlType(SqlTypeName.INTEGER);
+      return makeAbstractCast(type, makeZeroValue(i));
```

Comment:
The `makeAbstractCast` method with two entries has been marked as deprecated. You can use `makeAbstractCast(type, makeZeroValue(i), false)`.